bin/dry-run.rb

```diff
@@ -80,68 +81,120 @@
+  repo_name: nil,
+}
+
 if ENV["LOCAL_GITHUB_ACCESS_TOKEN"]
-  credentials << {
+  OPTIONS[:credentials] << {
     "type" => "git_source",
     "host" => "github.com",
     "username" => "x-access-token",
     "password" => ENV["LOCAL_GITHUB_ACCESS_TOKEN"]
   }
 end
 
-# Directory where the base dependency files are.
-directory = "/"
-
-# Name of an individual dependency to udpate
-dependency_name = nil
+if ENV["LOCAL_CONFIG_VARIABLES"]
+  # For example:
+  # LOCAL_CONFIG_VARIABLES="[{\"type\":\"npm_registry\",\"registry\":\"registry.npmjs.org\",\"token\":\"123\"}]"
+  OPTIONS[:credentials].concat(JSON.parse(ENV["LOCAL_CONFIG_VARIABLES"]))
+end
 
 option_parse = OptionParser.new do |opts|
   opts.banner = "usage: ruby bin/dry-run.rb [options] PACKAGE_MANAGER REPO"
 
   opts.on("--dir DIRECTORY", "Dependency file directory") do |value|
-    directory = value
+    OPTIONS[:directory] = value
+  end
+
+  opts.on("--branch BRANCH", "Repo branch") do |value|
+    OPTIONS[:branch] = value
   end
 
   opts.on("--dep DEPENDENCY", "Dependency to update") do |value|
-    dependency_name = value
+    OPTIONS[:dependency_name] = value
+  end
+
+  opts.on("--cache STEPS", "Cache e.g. files, dependencies, updates") do |value|
+    OPTIONS[:cache_steps].concat(value.split(",").map(&:strip))
   end
 end
+
 option_parse.parse!
 
-# Full name of the GitHub repo you want to create pull requests for.
+# Full name of the GitHub repo you want to create pull requests for
 if ARGV.length < 2
   puts option_parse.help
   exit 1
 end
-package_manager, repo_name = ARGV
+
+OPTIONS[:package_manager] = ARGV[0]
+OPTIONS[:repo_name] = ARGV[1]
 
 source = Dependabot::Source.new(
   provider: "github",
-  repo: repo_name,
-  directory: directory,
-  branch: nil
+  repo: OPTIONS[:repo_name],
+  directory: OPTIONS[:directory],
+  branch: OPTIONS[:branch]
 )
 
 # Fetch the dependency files
 puts "=> fetching dependency files"
-fetcher = Dependabot::FileFetchers.for_package_manager(package_manager).
-          new(source: source, credentials: credentials)
-files = fetcher.files
+
+fetcher = Dependabot::FileFetchers.for_package_manager(OPTIONS[:package_manager]).
+          new(source: source, credentials: OPTIONS[:credentials])
+
+def cache_read(name)
+  raise "Provide something to cache" unless block_given?
+  return yield unless OPTIONS[:cache_steps].include?(name)
+
+  repo_path = OPTIONS[:repo_name].split("/").join("@")
+  cache_path = File.join("tmp", repo_path, "cache", "#{name}.bin")
+  cache_dir = File.dirname(cache_path)
+  FileUtils.mkdir_p(cache_dir) unless Dir.exist?(cache_dir)
+  cached = File.read(cache_path) if File.exist?(cache_path)
+
+  if cached
+    Marshal.load(cached)
+  else
+    data = yield
+    File.write(cache_path, Marshal.dump(data))
+    data
+  end
+end
+
+files = cache_read("files") { fetcher.files }
+
+# Dump dependency files in tmp/githublogin@repo-name/dependency-files
+files.map do |f|
+  repo_path = OPTIONS[:repo_name].split("/").join("@")
```

Comment:
Again, maybe remove the "@" and just use another directory?

```suggestion
  repo_path = OPTIONS[:repo_name]
```